paf-mvp-client-express/src/origin-validator.ts

```diff
@@ -0,0 +1,107 @@
+import cors, { CorsRequest } from 'cors';
+import { NextFunction, Request, Response } from 'express';
+import { Log } from '@core/log';
+import { escapeRegExp, getTopLevelDomain } from '@core/express';
+import { ClientNodeError, ClientNodeErrorType } from '@core/errors';
+import { proxyUriParams } from '@core/endpoints';
+
+/**
+ * Class to manipulate a allowed origins and referrers based on current hostname
+ */
+export class OriginValidator {
+  private readonly allowedOrigins: RegExp[];
+  public cors: (req: CorsRequest, res: Response, next: (err?: unknown) => unknown) => void;
+
+  constructor(protected hostName: string, protected logger: Log) {
+    const tld = getTopLevelDomain(hostName);
+
+    // Only allow calls from the same TLD+1, under HTTPS
+    this.allowedOrigins = [
+      new RegExp(
+        `^https:\\/\\/(([a-zA-Z0-9]|[a-zA-Z0-9][a-zA-Z0-9\\-]*[a-zA-Z0-9])\\.)*${escapeRegExp(tld)}(/?$|\\/.*$)`
+      ),
+    ];
+
+    const corsOptions = {
+      origin: this.allowedOrigins,
+      optionsSuccessStatus: 200, // some legacy browsers (IE11, various SmartTVs) choke on 204
+      credentials: true,
+      allowedHeaders: ['Content-Type'],
+    };
+
+    this.cors = cors(corsOptions);
+  }
+
+  private isValidOrigin(origin: string) {
+    return this.allowedOrigins.findIndex((regexp: RegExp) => regexp.test(origin)) !== -1;
+  }
+
+  /**
+   * Returns a handler that will verify the provided origin is valid
+   * @param endpoint
+   */
+  checkOrigin(endpoint: string) {
```

Comment:
You may not need this parameter. Use `req.path`.


````typescript
checkOrigin() {
    return (req: Request, res: Response, next: NextFunction) => {
        const endpoint = req.path;
        // ....
    }
}
````


Since you don't need a parameter, I think you have two options: 
1. keeping the method that returns a function and keep the `this`
2. Make a simple function that you bind.

I would be OK for the second one because we already bind some method in the client-node.ts 
Something like this ?
```typescript
checkOriginHandler(req: Request, res: Response, next: NextFunction) => {
        const endpoint = req.path;
        // ....
}
// client-node.ts
this.app.expressApp.get(
    jsonProxyEndpoints.read,
    this.originValidator.cors,
    this.originValidator.checkOrigin.bind(this.originValidator),
    this.startSpan(jsonProxyEndpoints.read),
    this.restRead.bind(this),
    this.handleErrors(jsonProxyEndpoints.read),
    this.endSpan(jsonProxyEndpoints.read)
);
```

Note that this suggestion can be applied to other public methods of this class but also to the methods in `node.ts`

Reply:
Actually, the logging (the only reason for passing the endpoint route) can be removed as now the errors are logged directly by the `handleErrors` method.
I have removed it in a new commit.
However, in the case where something is logged, I foresee that we won't use just the endpoint _route_, but probably an endpoint _name_, which cannot be part of the request parameter. So I prefer to leave the `endpoint` parameter as it is for the moment.

Reply:
As for the `.bind(this)`, thanks for mentioning it, I wasn't happy about it 😉
So I got rid of it, replacing methods with arrow functions (which carry the "this" context).

Reply:
What would be your endpoint name for instance? What would be the value of it?

The drawback of such an approach is the creation of a level of indirection and so complexity.

Reply:
I don't know yet what the name would be, it is going to be driven by monitoring.
In any case, the route won't be enough: since we are using REST, the same route has different logic depending on the HTTP method.
And we are always free to remove it later if we decide, for instance, that "HTTP method + route" is the right "name".

Reply:
Sorry, I did not push my commits 🤦‍♂️
On the last version I think you will see this indirection doesn't exist anymore

Reply:
> I don't know yet what the name would be, it is going to be driven by monitoring.
> In any case, the route won't be enough: since we are using REST, the same route has different logic depending on the HTTP method.
> And we are always free to remove it later if we decide, for instance, that "HTTP method + route" is the right "name".

Yes, we should use the endpoint, not the route (endpoint = route + verb).

Up to you. I would use a YAGNI approach instead.

Reply:
This will be decided when implementing the monitoring metrics